### PR TITLE
Wrap objects before adding them to `AnnotatedQueue`

### PR DESCRIPTION
**Pull request by mudit2812**
*Description of the Change:**
* Create `WrappedObj` class that wraps objects and implements `__hash__` and `__eq__` that is independent of the object's `__hash__` and `__eq__`.
* Updated `AnnotatedQueue` to wrap all queued objects in `WrappedObj` and update class logic accordingly
* Update `QuantumScript`, `QuantumTape` to work with the wrapped objects.
* Minor documentation and helper function fixes that were missed when shots were added to `QuantumScript`.

**Benefits:**
Once operators are immutable, `Operator.__eq__` can be overridden to use `qml.equal` as queuing will no longer depend on the `Operator`'s hash.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/tape/tape.py [hunk 1]
```diff
@@ -184,7 +184,8 @@ def stop_at(obj):  # pylint: disable=unused-argument
         (tape._ops + diagonalizing_gates, new_ops),
         (diagonal_measurements, new_measurements),
     ]:
-        for obj in queue:
+        for wrapped_obj in queue:
+            obj = wrapped_obj.obj
```
Comment (albi3ro) [inline pennylane/tape/tape.py:188]: At this point, shouldn't we be dealing with the objects themselves?

Comment (mudit2812) [reply pennylane/tape/tape.py:188]: Thanks for pointing that out :)

**Comment by mudit2812** [comment]
[sc-36544]

File: pennylane/queuing.py [hunk 2]
```diff
@@ -357,29 +373,50 @@ def __exit__(self, exception_type, exception_value, traceback):
...
         return self[obj]
 
+    def items(self):
+        items = super().items()
+        new_items = []
+
+        for key, value in items:
+            new_items.append((key.obj, value))
+
+        return tuple(new_items)
```
Comment (albi3ro) [inline pennylane/queuing.py:406]:
```suggestion
        return tuple((key.obj, value) for key, value in super().items())
```
should work.

File: pennylane/tape/qscript.py [hunk 3]
```diff
@@ -1045,7 +1045,9 @@ def copy(self, copy_operations=False):
             _ops = self._ops.copy()
             _measurements = self.measurements.copy()
 
-        new_qscript = self.__class__(ops=_ops, measurements=_measurements, prep=_prep)
+        new_qscript = self.__class__(
+            ops=_ops, measurements=_measurements, prep=_prep, shots=self.shots
+        )
```
Comment (albi3ro) [inline pennylane/tape/qscript.py:1050]: Should this go into a separate PR?

File: pennylane/queuing.py [hunk 4]
```diff
@@ -173,6 +173,22 @@ class QueuingError(Exception):
...
+        self.obj = obj
+
+    def __hash__(self):
+        return id(self.obj)
+
+    def __eq__(self, other):
+        return id(self.obj) == id(other.obj)
+
+    def __repr__(self):
+        return self.obj.__repr__()
```
Comment (albi3ro) [inline pennylane/queuing.py:189]: Probably best to differentiate between and object and its wrapped version in the repr. Otherwise, things could get hard to debug:
```suggestion
        return f"Wrapped( {self.obj.__repr__()})"
```

Comment (mudit2812) [reply pennylane/queuing.py:189]: Good idea

**Review by albi3ro — COMMENTED** [review]
We should break the shot changes into a separate PR.

The second comment is one about tests.

Right now, we can't really unit test that `AnnotatedQueue` doesn't depend on an objects hashability, because `Operator` gives exactly the same behaviour before and after this change.
But we could write some unit tests for a different object to make sure that `AnnotatedQueue` doesn't depend on its hash.

For example:
```
q = qml.queuing.AnnotatedQueue()

a = [1,"b"]
q.append(a)
q.append({"a": "c"})
q.queue
```

**Review by albi3ro — APPROVED** [review]
Thanks for this. Glad to finally be making progress on these behaviors and assumptions :)

File: pennylane/queuing.py [hunk 5]
```diff
@@ -173,6 +173,22 @@ class QueuingError(Exception):
...
+    """Wraps an object to make its has dependent on its identity"""
+
+    def __init__(self, obj):
+        self.obj = obj
+
+    def __hash__(self):
+        return id(self.obj)
+
+    def __eq__(self, other):
+        return id(self.obj) == id(other.obj)
```
Comment (Jaybsoni) [inline pennylane/queuing.py:186]: Is there any advantage to storing the id as an attribute? This way we don't have to call `hash` multiple times?

Comment (mudit2812) [reply pennylane/queuing.py:186]: I don't think the speedup would be significant enough to matter, and storing the id as an attribute makes the code a bit less readable in my opinion.

**Review by Jaybsoni — APPROVED** [review]
Looks good! Just wondering if there would be any issues if we update the object (say update its parameters or map its wires for example) with the hash and its place in the queue.

File: pennylane/queuing.py (hunk 5 above)
Comment (Jaybsoni) [reply pennylane/queuing.py:186]: Sounds good 👍🏼

File: pennylane/queuing.py [hunk 6]
```diff
@@ -173,6 +173,22 @@ class QueuingError(Exception):
     """Exception that is raised when there is a queuing error"""
 
 
+class WrappedObj:
+    """Wraps an object to make its has dependent on its identity"""
```
Comment (mudit2812) [inline pennylane/queuing.py:177]:
```suggestion
    """Wraps an object to make its hash dependent on its identity"""
```

**Comment by mudit2812** [comment]
@Jaybsoni That's a good point. Afaik the id of a Pennylane object shouldn't change if you update it so I don't think (emphasis on think) that this would be a problem. That being said, operator immutability is something that is on the roadmap so this would definitely not be an issue in the future.
